src/serlio/prtMaterial/ArnoldMaterialNode.cpp

```diff
@@ -92,147 +105,97 @@ MStatus ArnoldMaterialNode::compute(const MPlug& plug, MDataBlock& data) {
+	scriptBuilder.declString(L"$shadingGroup");
+	scriptBuilder.declString(L"$shaderNode");
+	scriptBuilder.declString(L"$mapFile");
+	scriptBuilder.declString(L"$mapNode");
+	scriptBuilder.declString(L"$bumpLuminanceNode");
+	scriptBuilder.declString(L"$bumpValueNode");
+	scriptBuilder.declString(L"$displacementNode");
+	scriptBuilder.declString(L"$normalMapConvertNode");
+	scriptBuilder.declString(L"$colorMapBlendNode");
+	scriptBuilder.declString(L"$dirtMapBlendNode");
+	scriptBuilder.declString(L"$opacityMapBlendNode");
+	scriptBuilder.declString(L"$specularMapBlendNode");
+	scriptBuilder.declString(L"$emissiveMapBlendNode");
+	scriptBuilder.declString(L"$roughnessMapBlendNode");
+	scriptBuilder.declString(L"$metallicMapBlendNode");
+	scriptBuilder.declString(L"$uvTrafoNode");
 
 	for (adsk::Data::Handle& inMatStreamHandle : *inMatStream) {
 		if (!inMatStreamHandle.hasData())
 			continue;
 
-		if (!inMatStreamHandle.usesStructure(*fStructure))
+		if (!inMatStreamHandle.usesStructure(*materialStructure))
 			continue;
 
-		MaterialInfo matInfo(inMatStreamHandle);
-
-		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceStart.c_str())) {
+		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceStart.c_str()))
 			continue;
-		}
 		const int faceStart = *inMatStreamHandle.asInt32();
 
-		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceEnd.c_str())) {
+		if (!inMatStreamHandle.setPositionByMemberName(gPRTMatMemberFaceEnd.c_str()))
 			continue;
-		}
 		const int faceEnd = *inMatStreamHandle.asInt32();
 
-		// get unique name
-		std::wstring shaderName(L"serlioGeneratedArnoldMaterialSh");
-		std::wstring shadingGroupName(L"serlioGeneratedArnoldMaterialSg");
+		auto createShadingEngine = [this, &nameSource, &materialStructure, &scriptBuilder,
+		                            &inMatStreamHandle](const MaterialInfo& matInfo) {
+			auto newMaterialName = nameSource.getUniqueName(MATERIAL_BASE_NAME);
+			const std::wstring& shadingEngineName = newMaterialName.first;
+			const std::wstring& shaderName = newMaterialName.second;
+
+			synchronouslyCreateShadingEngine(shadingEngineName);
+			MaterialUtils::assignMaterialMetadata(materialStructure, inMatStreamHandle, shadingEngineName);
+			appendToMaterialScriptBuilder(scriptBuilder, matInfo, shaderName, shadingEngineName);
+			LOG_INF << "new arnold shading engine: " << shadingEngineName << " (surface shader: " << shaderName << ")";
 
-		int shIdx = 0;
-		while (shaderNames.find(shaderName + std::to_wstring(shIdx)) != shaderNames.end()) {
-			shIdx++;
-		}
-		shaderName += std::to_wstring(shIdx);
-		shadingGroupName += std::to_wstring(shIdx);
-		shaderNames.insert(shaderName);
+			return shadingEngineName;
+		};
 
-		buildMaterialShaderScript(sb, matInfo, shaderName, shadingGroupName, meshName.asWChar(), faceStart, faceEnd);
+		MaterialInfo matInfo(inMatStreamHandle);
+		std::wstring shadingEngineName = getCachedValue(matCache, matInfo, createShadingEngine, matInfo);

```

Comment:
it wouldn't be necessary to add the shading engine name to the cache in `getCachedValue` in case a new one gets created, since the cache is never used afterwards anymore

Reply:
this is for the case a mesh has multiple materials

Reply:
oh true, my bad, overlooked the surrounding loop, all good then